pkg/engine/lifeycletest/refresh_test.go

```diff
@@ -813,3 +813,49 @@ func TestRefreshStepWillPersistUpdatedIDs(t *testing.T) {
 		}
 	}
 }
+
+// TestRefreshUpdateWithDeletedResource validates that the engine handles a deleted resource without error on an
+// update with refresh.
+func TestRefreshUpdateWithDeletedResource(t *testing.T) {
```

Comment:
We might need to add an analyzer to this test in order to properly validate the originally failing codepath. I don't think we've done that before, but it should be doable. LMK if you'd like some pointers on where to start.

Reply:
This test panics without the fix applied, but not 100% sure if it's the same spot.

Reply:
Ok, confirmed that the test panics in the same spot, so this should be good to go.

```
=== RUN   TestRefreshUpdateWithDeletedResource
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0x19f6710]

goroutine 55 [running]:
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepGenerator).AnalyzeResources.func1(0xc000270f00, 0x35, 0xc0004e4480, 0xc0003d7600)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/step_generator.go:1301 +0x130
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*resourceMap).mapRange.func1(0x25347a0, 0xc0002491e0, 0x2405ac0, 0xc0004e4480, 0x1)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/deployment.go:127 +0x6b
sync.(*Map).Range(0xc000591200, 0xc0006a7810)
	/usr/local/Cellar/go/1.15.5/libexec/src/sync/map.go:345 +0x1d9
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*resourceMap).mapRange(0xc000591200, 0xc0006a7970)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/deployment.go:126 +0x55
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepGenerator).AnalyzeResources(0xc0004e4a20, 0x0, 0x0)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/step_generator.go:1289 +0xac
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*deploymentExecutor).Execute(0xc00000e820, 0x2938e60, 0xc000591290, 0x9c90e58, 0xc0000e2c80, 0x0, 0x1, 0x0, 0x0, 0x0, ...)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/deployment_executor.go:277 +0xae6
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*Deployment).Execute(...)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/deployment.go:384
github.com/pulumi/pulumi/pkg/v2/engine.(*deployment).run.func1(0x293dd20, 0xc0000e2c80, 0xc0000e2000, 0x2938e60, 0xc000591290, 0xc000591201, 0xc000248ec0, 0xc000389320)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/engine/deployment.go:252 +0x287
created by github.com/pulumi/pulumi/pkg/v2/engine.(*deployment).run
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/engine/deployment.go:237 +0x306
```